controllers/topics/broadcastMessage.js

```diff
@@ -0,0 +1,117 @@
+  };
+  try {
+    await followTopicQueue.add(data, options);
+  } catch (e) {
+    console.log('error', e);
+  }
+};
+
+const calculateDelay = (joinDate, delay) => {
+  let currentTime = momentTz().tz('America/Los_Angeles');
+  let requiredTime = momentTz(joinDate).tz('America/Los_Angeles');
+
+  let diffTime = requiredTime.diff(currentTime, 'days');
+  diffTime = Math.abs(diffTime + delay);
+  return diffTime == 0 ? 60 * 1000 : diffTime * 24 * 60 * 60 * 1000;
+};
+
+module.exports = async (req, res) => {
+  const {userId} = req;
+  const {communityMessageFormatId, topicId, targetUserId, targetDate, delay} = req.body;
+
+  //   check if the the user is anonymous
+  const detailTokenUser = await UsersFunction.findUserById(User, userId);
+  if (!detailTokenUser) {
+    return res.status(404).json({message: 'User not found'});
+  }
+  if (detailTokenUser.is_anonymous) {
+    return res.status(403).json({message: 'Anonymous user cannot send message'});
+  }
+  // find community message format if defined
+  if (communityMessageFormatId) {
+    const communityMessageFormat = await CommunityMessageFormat.findByPk(communityMessageFormatId);
+    if (!communityMessageFormat) {
+      return res.status(404).json({message: 'Community message format not found'});
+    }
+    sendMessageToQueue(
+      targetUserId,
+      communityMessageFormat.topic_id,
+      communityMessageFormatId,
+      delay || communityMessageFormat.delay
+    );
+  } else if (topicId) {
+    try {
+      const communityMessageFormats = await CommunityMessageFormat.findAll({
+        where: {topic_id: topicId, status: '1'}
+      });
+      if (communityMessageFormats && !detailTokenUser.is_anonymous) {
+        for (const format of communityMessageFormats) {
+          if (targetUserId) {
+            sendMessageToQueue(
+              targetUserId,
+              topicId,
+              format.community_message_format_id,
+              delay || format.delay
+            );
+          } else {
+            // get all user who follow the topic in certain period
+            const users = await UserTopic.findAll({
+              where: {
+                topic_id: topicId,
+                createdAt: {
+                  [Op.between]: [
+                    momentTz(targetDate).startOf('day').toDate(),
+                    momentTz(targetDate).endOf('day').toDate()
+                  ]
+                }
+              }
+            });
+            for (const user of users) {
+              // calculate the delay time
+              let calculatedDelay = calculateDelay('2024-09-02', format.delay_time);
+              sendMessageToQueue(
+                user.user_id,
+                topicId,
+                format.community_message_format_id,
+                calculatedDelay
+              );
+            }
+          }
+        }
+      }
+    } catch (error) {
+      console.log(error);
+      return res.status(404).json({message: 'Error while sending message'});
+    }
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

Similar to the previous comment, error handling here could be improved. Instead of just logging the error and returning a generic message, consider providing more context about the error to the client (if it's safe to do so) or at least log more detailed information for debugging purposes.
```diff
-    console.log(error);
-    return res.status(404).json({message: 'Error while sending message'});
+    console.error(`Error while sending message: ${error.message}`);
+    return res.status(500).json({message: `Error while sending message: ${error.message}`});
```



<!-- This is an auto-generated comment by OSS CodeRabbit -->